bazel: build panda proxy and schema registry

dotnwat opened the pull request:
Build panda proxy and schema registry with bazel.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

dotnwat commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -17,15 +17,32 @@
 #include "pandaproxy/logger.h"
 #include "pandaproxy/schema_registry/errors.h"
 #include "pandaproxy/schema_registry/sharded_store.h"
+#include "src/v//pandaproxy/schema_registry/protobuf/confluent/meta.pb.h"
```
Improving the situation here is captured in this ticket. https://redpandadata.atlassian.net/browse/CORE-6793. On the plus side, these are included on in a single compilation unit, and we've altered the include path only for this specific file in cmake.

dotnwat replied:
Shared issue with https://github.com/redpanda-data/redpanda/pull/22714

rockwotj replied:
Why is it `src/v//pandaproxy` instead of `src/v/pandaproxy`?

dotnwat replied:
It's a typo, I'll fix it. I guess the preprocessor don't care

dotnwat replied:
fixed

dotnwat commented on src/v/cluster/BUILD:
```diff
@@ -564,8 +564,8 @@ redpanda_cc_library(
         "//src/v/net",
-        "//src/v/pandaproxy/schema_registry:config",
-        "//src/v/pandaproxy/schema_registry:subject_name_strategy",
+        "//src/v/pandaproxy:config",
+        "//src/v/pandaproxy:subject_name_strategy",
```
these are re-rooted so that panda proxy and schema registry can be compiled together as a monolithic library. not ideal, but I encountered various violations of module boundaries and I don't understand pp/sr well enough to fix these efficiently. will rely on ben/oren to help break this down as follow up work.

dotnwat commented on src/v/pandaproxy/BUILD:
```diff
@@ -0,0 +1,336 @@
+    ],
+)
+
+cc_proto_library(
+    name = "field_mask_cc_proto",
+    deps = [
+        "@protobuf//:field_mask_proto",
+    ],
+)
```
this was required to resolve layering check violations.

dotnwat commented on src/v/thirdparty/protobuf/BUILD:
```diff
@@ -0,0 +1,16 @@
+# interfaces. The `.h` file is not listed anywhere as a public header, and the
+# `.pb.h` beats me. all of the other proto files work as expected.
+#
+cc_library(
+    name = "descriptor",
+    hdrs = [
+        "descriptor.h",
+        "descriptor.pb.h",
```
Fixes for layering check violations. This one is still a bit mysterious to me. Well, for `descriptor.h` it's straight forward: in v24 this doesn't seem to be exported through a publicly visible library. For the `pb.h` one it's weird. Even if I create an explicit proto_cc_library for it, it doesn't seem to see it.

In any case, none of this is necessary when we move to protobuf v26, but i ran into other compat issues there as we need to re-implement some bits that rely on fully deprecated/removed interfaces.

rockwotj replied:
I'm going to have to upgrade to the latest protobuf for the protobuf parser so I can dig into this.

dotnwat replied:
>I'm going to have to upgrade to the latest protobuf for the protobuf parser so I can dig into this.

Amazing. I briefly looked at the the replacement interface for the deprecated bits and it didn't look too cray.

BenPope replied:
https://github.com/BenPope/redpanda/tree/update_libprotobuf

I didn't have time to look at disabling the logging.

rockwotj replied:
https://github.com/protocolbuffers/protobuf/issues/16914

BenPope replied:
> [protocolbuffers/protobuf#16914](https://github.com/protocolbuffers/protobuf/issues/16914)

Yeah, that's where I stopped.

dotnwat commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -241,7 +241,8 @@ class parser {
         }
-        _fdp.set_name(schema.sub()());
+        const auto& sub = schema.sub()();
+        _fdp.set_name(sub.data(), sub.size());
```
std::string_view also works here in bazel, but protobuf used in cmake oss build on fc38 doesn't seem to have this overload.

BenPope replied:
Oh, the vtools build seems this is enabled:
```c++
#ifdef ABSL_USES_STD_STRING_VIEW

#include <string_view>  // IWYU pragma: export

namespace absl {
ABSL_NAMESPACE_BEGIN
using string_view = std::string_view;
ABSL_NAMESPACE_END
}  // namespace absl
```

dotnwat replied:
Ah, good spot. tracking here it's gonna take a bi tof investigation to figure out how to get abseil config chagnees to propogate through its depdency rather than doing something like setting this define project wide. not sure how that's done with bzlmod yet. https://redpandadata.atlassian.net/browse/CORE-6800

rockwotj replied:
Using `std::string_view` for `absl::string_view` is enabled for the bazel build too?

https://github.com/abseil/abseil-cpp/blob/08850701d27544b4af6501154791d3d45e28b0f8/absl/base/config.h#L567

The issue is 24.4 is an old version of proto and doesn't have a string view overload. It doesn't have anything to do with absl?

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/52494#019124c0-3d97-419b-8213-3bd820428f66
<!--76b333d4-6994-408a-96c8-ba57ab25af9c-->

BenPope reviewed (commented):
Looks good to me, the protobuf deps looked fun to figure out.

dotnwat commented on src/v/pandaproxy/error.cc:
```diff
@@ -12,6 +12,7 @@
 #include "error.h"
 
 #include "kafka/protocol/errors.h"
+#include "pandaproxy/error.h"
```
ah yeh at this point error.h and pandaproxy/error.h are probably referring to the same file.

dotnwat replied:
fixed

dotnwat commented:
Force push https://github.com/redpanda-data/redpanda/compare/cd6696f19f19baad274a59767c14d44c82ca1029..0ee002bf3d149947359d80038e4901a3747bb24d
* fixed header include path for protobuf headers
* fixed redundant error.h inclusion

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/52548#019128e8-63c7-49fe-8f55-89c47003138d:
```
"rptest.tests.e2e_iam_role_test.ShortLivedCredentialsTests.test_short_lived_credentials"
```
<!--a4a4be4b-17d7-41e4-8058-61bb1d867385-->
